src/io/filewatcher.rs

```diff
@@ -4,120 +4,83 @@
-                let last = file.metadata().unwrap().modified().ok();
-                Ok(FileWatcher {
-                    file_path: path.into_boxed_path(),
-                    cpos: pos,
-                    lines_buf: VecDeque::new(),
-                    last_read: last,
-                })
-            }
-            Err(err) => Err(err),
-        }
+    fn open_file<P: AsRef<Path>>(filepath: P) -> Result<File> {
+        File::open(filepath).context("Failed to open log file")
     }
 
-    fn open_file(filepath: PathBuf) -> Result<File, io::Error> {
-        match File::open(&filepath) {
-            Ok(x) => Ok(x),
-            Err(err) => {
-                if let Ok(path) = filepath.into_os_string().into_string() {
-                    log::error!("Failed to open log file {}: {}", path, err);
-                } else {
-                    log::error!("Failed to open log file: {}", err);
-                }
-                Err(err)
-            }
-        }
-    }
+    /// Attempts to read the new contents of the observed file and updates the internal state
+    /// with any new lines that have been appended since last call.
+    fn read_new_file_lines(&mut self) -> Result<()> {
+        let meta =
+            std::fs::metadata(&self.file_path).context("Failed to fetch metadata for log file.")?;
 
-    fn seek_to_pos(&self) -> Option<ReadMD> {
-        if let Ok(mut file) = FileWatcher::open_file(self.file_path.clone().into_path_buf()) {
-            if let Ok(modified) = file.metadata().unwrap().modified() {
-                if let Some(last_read) = self.last_read {
-                    if modified <= last_read {
-                        return None;
-                    }
-                } else {
-                    log::error!("No stored last mod time.");
-                }
-            } else {
-                log::error!("Unable to read modification time of watched file.");
-            }
-            let mut buff: Vec<u8> = Vec::new();
-            let _ = file.read_to_end(&mut buff);
+        // No new data
+        if meta.len() as usize == self.last_size || meta.len() == 0 {
+            return Ok(());
+        }
 
-            let mut reader = Cursor::new(buff);
-            reader.set_position(self.cpos);
+        // Get new file contents
+        let mut file =
+            FileWatcher::open_file(&self.file_path).context("Failed to reopen log file")?;
+        let mut buff: Vec<u8> = Vec::new();
+        let _ = file.read_to_end(&mut buff);
+        let mut start_idx = self.last_size;
 
-            if let Ok(mod_time) = file.metadata().unwrap().modified() {
-                Some(ReadMD(Some(mod_time), reader))
-            } else {
-                Some(ReadMD(None, reader)) // TODO: fix this grossness
-            }
-        } else {
-            None
+        // Reset if file has been remade (i.e. is shorter) and update state
+        if buff.len() < self.last_size {
```

Comment:
I think this should be before line 51 where we get the new contents, and probably should check that `meta.len()` is less than the last size. If the file has reset but we are asking for new data, we are likely going to want whatever is sitting in the front of that file immediately. Thoughts?

Reply:
If I'm understanding you correctly, that is basically what is already happening as `buff.len() == meta.len()`. Although yeah if it's more clear to use `meta.len()` and do that beforehand then I'll move it there.

Reply:
i mean yeah semantically its the same, but i see how you're actually using that variable. I hadn't exactly noticed you'd removed the cursor seek. this is no different behavioru then